File: agent/video_agent/prompts.py
"""LLM 系统 prompt 与消息构建。"""

SYSTEM_PROMPT = r'''你是一名资深的动态视频工程师,负责为 "mini-remotion"(一个类 Remotion 的、基于 React 帧驱动的视频引擎)编写视频组件。

# 你的输出契约(必须严格遵守)
- 只输出**一个** ```tsx 代码块,不要有任何额外解释文字、不要有多个代码块。
- 该文件必须从 "../core" 导入所需 API(**只能**从 "../core" 导入,禁止其它任何 import)。
- 必须导出:
  1. `export const meta = { width, height, fps, durationInFrames }`(纯数字字面量)。
  2. `export const VideoComposition: React.FC`(无 props)。
- 也需要 `import React from "react";`。

# 可用 API(全部来自 "../core")
- `useCurrentFrame(): number` —— 当前帧号(在 <Sequence> 内是相对帧号)。这是唯一的时间来源。
- `interpolate(frame, inputRange, outputRange, options?)` —— 线性映射。
- `spring({ frame, fps, from?, to?, config? })` —— 弹簧动画。
- `Easing` —— { linear, easeIn, easeOut, easeInOut }。
- `random(seed: number | string): number` —— 确定性随机。**禁止 Math.random()**。
- `<Sequence from={n} durationInFrames={n}>...</Sequence>` —— 时间轴平移。
- `<Img src={...} />`、`staticFile("name.ext")`、`<Audio src={...} volume? />`。

# 确定性规则
- 画面必须是 useCurrentFrame() 的纯函数。
- 禁止 Math.random()、Date.now()、new Date()、setTimeout、setInterval、fetch、window/document。
- 不要使用 CSS 动画 / transition / requestAnimationFrame。

# 视觉规范
- 根元素 `style={{ position: "absolute", inset: 0 }}` 铺满。
- 全部内联 style,system-ui 字体,1280x720 @ 30fps,时长 90~240 帧。
'''


def build_repair_message(previous_code: str, errors: str) -> dict:
    return {
        "role": "user",
        "content": f"""你上一版生成的组件校验失败。请修复所有问题后,**重新输出完整的单个 tsx 代码块**。

## 上一版代码
```tsx
{previous_code}
```

## 校验错误
{errors}""",
    }


def enrich_prompt_with_tts(
    prompt: str, mp3_path: str, duration_seconds: float, fps: int = 30
) -> str:
    frames = int(duration_seconds * fps + 0.999)
    return prompt + f"""

【配音要求】
- 旁白已生成:public/{mp3_path},时长约 {duration_seconds:.1f}s。
- 请在组件中使用 <Audio src={{staticFile("{mp3_path}")}} volume={{0.9}} />。
- 将 meta.durationInFrames 设为 {frames}(按 {fps}fps 与旁白对齐)。"""

File: harness/prompts.ts
import type { ChatMessage } from "./llm/types.js";

export const SYSTEM_PROMPT = `你是一名资深的动态视频工程师,负责为 "mini-remotion"(类 Remotion 的 React 帧驱动视频引擎)编写视频组件。

# 输出契约
- 只输出**一个** \`\`\`tsx 代码块,不要解释文字。
- 只能从 "react" 和 "../core" 导入。
- 必须导出 \`export const meta = { width, height, fps, durationInFrames }\` 和 \`export const VideoComposition: React.FC\`。

# 可用 API(来自 "../core")
useCurrentFrame, interpolate, spring, Easing, random, Sequence, Img, staticFile, Audio

# 确定性规则
- 禁止 Math.random(), Date.now(), setTimeout, fetch, CSS 动画。
- 动画只依赖 useCurrentFrame()。

# 规范
1280x720 @ 30fps, durationInFrames 90~240, 内联 style, system-ui 字体。`;

export const buildUserMessage = (prompt: string): ChatMessage => ({
  role: "user",
  content: prompt,
});

export const buildRepairMessage = (
  previousCode: string,
  errors: string,
): ChatMessage => ({
  role: "user",
  content: `校验失败,请修复并重新输出完整 tsx 代码块。

## 上一版
\`\`\`tsx
${previousCode}
\`\`\`

## 错误
${errors}`,
});

export const enrichPromptWithTts = (
  prompt: string,
  mp3Path: string,
  durationSeconds: number,
  fps = 30,
): string => {
  const frames = Math.ceil(durationSeconds * fps);
  return `${prompt}

【配音要求】
- 旁白: public/${mp3Path}, 时长 ${durationSeconds.toFixed(1)}s
- 使用 <Audio src={staticFile("${mp3Path}")} volume={0.9} />
- meta.durationInFrames = ${frames}`;
};
